whole-tale/girder_wholetale
# Add notifications for shared events

**Problem**:
Under a number of scenarios, the dashboard must be manually refreshed to reflect changes to underlying data, particularly when a tale is shared with another user (see https://github.com/whole-tale/ngx-dashboard/issues/121).

**Approach:**
This PR adds support for sending specific Girder events to users with access to a Tale for the following:
* Tale is created (`wt_tale_created`), updated (`wt_tale_updated`),  (`wt_tale_removed`)
* Tale is shared (`wt_tale_shared`) or unshared with a user (`wt_tale_unshared`)
* Import is started (`wt_import_started`), completed (`wt_import_completed`) or failed (`wt_import_failed`)
* Instance is started (`wt_instance_launching`), running (`wt_instance_running`), or in error state (`wt_instance_error`)
This PR adds support for sending specific Girder events to users with access to a Tale for the following: Tale is shared with a user, Tale is unshared with a user, Tale is deleted, Tale is updated. This PR also adds progress event notifications for import jobs (should be another PR).

All events have the following format:
```
{
 '_id': '60395380822a8592ebb53562',
 'data': {
   'event': 'wt_tale_created',
   'modelType': 'tale',
   'resource': '6039537f822a8592ebb53560',
   'resourceName': 'WT event'
 },
 'expires': '2021-02-26T21:01:04.015000+00:00',
 'startTime': 1614369664.0172398,
 'time': '2021-02-26T20:01:04.017000+00:00',
 'type': 'wt_event',
 'updated': '2021-02-26T20:01:04.017000+00:00',
 'updatedTime': 1614369664.0172398,
 'userId': '6039537a822a8592ebb5354d'
}
```

Where:
* `type` is always `wt_event` (as a class of events to differentiate from `wt_progress`)
* `data.event` is the name of the specific event
* `modelType` is the type of resource (`tale` or `instance` with this PR)
* `resources` is the ID of the resource

**Future work:**
The following will be addressed in a separate/future PR:
* Events for file/folder updates
* Events for current jobs (build image, publish, log state)

**To Test:**
For each test step, use the `GET /notification` API to see the associated notifications
* Create a Tale (event `wt_tale_created`)
* Share the tale with another user (event `wt_tale_shared` send to other user)
* Unshare the tale with another user (event `wt_tale_unshared` send to other user)
* Share again then update metadata (event `wt_tale_updated` sent to both users)
* Start the instance (events `wt_instance_launching` and `wt_instance_running` to the current user only)
* Stop the instance (events `wt_instance_deleting` and `wt_instance_deleted` to the current user only)
* Delete the tale (event `wt_tale_deleted` sent to both users)
* Import a tale from Github, Zenodo, or as a Binder (events `wt_import_started`, `wt_import_complete`)
